### DISCUSSION TITLE
Blazor Static Server-Side Rendering (SSR) Integration with Custom OrchardCore Modules

### DISCUSSION THREAD

**Discussion by dturska** [OrchardCMS/OrchardCore#14951] [Q&A]
I am exploring the possibilities of integrating Blazor with OrchardCore, specifically using Blazor's (SSR) hosting model for developing custom OrchardCore modules. My interest lies in understanding the current capabilities and future plans regarding this integration.

Current State: As of now, is it possible to use Blazor Static SSR for building custom modules in OrchardCore? If so, are there any guides or examples available that demonstrate this integration?

Future Plans: Are there any ongoing developments or future plans to natively support Blazor, particularly the Static SSR hosting model, in OrchardCore? If such plans exist, could you provide any details or roadmap information?

![2023-12-24_11-27-10](https://github.com/OrchardCMS/OrchardCore/assets/1976670/e49378cc-cf22-4bf1-99e3-28441a5e17fd)

**Comment by ns8482e** [comment]
@dturska orchard core 1.7 works well with Blazor Server and WebAssembly.

Try OC 1.8 preview with .NET 8 that supports InteractiveAuto

**Comment by dturska** [comment]
Thank you for the reply.  I am trying to set up Blazor SSR as an OrchardCore Module with OC 1.8.0-preview-17894.  I am attempting to convert a Blazor Server-Side Rendering (SSR) project into a module for OrchardCore, specifically using version 1.8.0-preview-17894. I am looking for the appropriate way to configure an OrchardCore module—not the OrchardCore host—to incorporate Blazor SSR functionality. Any advice or instructions on this setup would be greatly appreciated.
![image](https://github.com/OrchardCMS/OrchardCore/assets/1976670/811f4de1-9073-4321-a7d8-4c1eb82b473d)

**Reply by ns8482e** [reply]
You can keep BlazorApp as separate project that contains all your `.razor` components. Add reference of your BlazorApp to your module project and configure Blazor services in startup in your module.

See discussion #14722

**Reply by dturska** [reply]
Thanks again for looking into this. 
I reviewed that discussion and I see that it addresses how to make Blazor WebAssembly work, but it doesn't seem to cover Blazor Server-Side Rendering (SSR). From what I've gathered from that discussion, there are two main projects involved: Zeugma and Zeugma.Client. Zeugma acts as the OrchardCore host and is responsible for configuring Blazor, while Zeugma.Client is a WebAssembly application. However, it appears that the Zeugma solution does not involve custom OrchardCore Modules that use Blazor SSR.

**Reply by ns8482e** [reply]
It covers both however, in your case you don't need to register services related to WebAssembly 
i.e. use only AddInteractiveServerComponents and   AddInteractiveServerRenderMode

**Reply by dturska** [reply]
I think I am going to address interactivity at another time.  I am focused on using the new Server Side Rendering feature of Blazor. 
I have updated my solution with an out of the box Blazor App using SSR. This project is referenced by an OrchardCore Module (BlazorAppOcModule).   Any  ideas on what the Startup.cs file should look like in the BlazorAppOcModule project? 
![2023-12-28_19-24-54](https://github.com/OrchardCMS/OrchardCore/assets/1976670/0439edc5-bc77-488c-a56f-b231476618da)

**Reply by ns8482e** [reply]
Answer is same as https://github.com/OrchardCMS/OrchardCore/discussions/14722#discussioncomment-7613709

 In your module's startup ConfigureServices add code for configure services 
 And on startup's Configure add code from Configure 

You don't need to register WebAssembly related services or skip registering services that you don't need

Additionally, You don't need Program.cs in BlazorApp, you can simply use RazorClassLibrary for Your BlazorApp project and add reference of that to your module's project

**Reply by dturska** [reply]
Thanks again. I got a bit further. The application now runs as an OrchardCore Module.  The css files aren't being located. One solution is to manually copy them to the BlazorAppOcModule wwwroot folder and then update the paths in the app.razor file. Not a great solution, but it works. 
![image](https://github.com/OrchardCMS/OrchardCore/assets/1976670/76394be6-a51a-4e80-835c-13c329c0767e)

**Reply by ns8482e** [reply]
Set base HREF in app.razor and set asset path in razorclasslib BlazorApp
